frontend/src/container/FormAlertRules/QuerySection.tsx

```diff
@@ -0,0 +1,297 @@
+			current.expression = expression;
+		}
+
+		if (toggleDisable) {
+			current.disabled = !current.disabled;
+		}
+
+		if (toggleDelete) {
+			delete allFormulas[formulaIndex];
+		} else {
+			allFormulas[formulaIndex] = current;
+		}
+
+		setFormulaQueries({
+			...allFormulas,
+		});
+	};
+
+	const handleMetricQueryChange = ({
+		queryIndex,
+		aggregateFunction,
+		metricName,
+		tagFilters,
+		groupBy,
+		toggleDisable,
+		toggleDelete,
+	}: IQueryBuilderQueryHandleChange): void => {
+		const allQueries = metricQueries;
+		const current = metricQueries[queryIndex];
+		if (aggregateFunction) {
+			current.aggregateOperator = aggregateFunction;
+		}
+		if (metricName) {
+			current.metricName = metricName;
+		}
+
+		if (tagFilters && current.tagFilters) {
+			current.tagFilters.items = tagFilters;
+		}
+
+		if (groupBy) {
+			current.groupBy = groupBy;
+		}
+
+		if (toggleDisable) {
+			current.disabled = !current.disabled;
+		}
+
+		if (toggleDelete) {
+			delete allQueries[queryIndex];
+		} else {
+			allQueries[queryIndex] = current;
+		}
+
+		setMetricQueries({
+			...allQueries,
+		});
+	};
+
+	const addMetricQuery = useCallback(async () => {
+		if (Object.keys(metricQueries).length > 4) {
+			notification.error({
+				message:
+					'Unable to create query. You can create at max 5 queries and 1 formulae.',
+			});
+			return;
+		}
+		const queryLabel = getNextQueryLabel();
+
+		const queries = metricQueries;
+		queries[queryLabel] = {
+			name: queryLabel,
+			queryName: queryLabel,
+			metricName: 'signoz_latency_count',
+			formulaOnly: false,
+			aggregateOperator: EAggregateOperator.NOOP,
+			legend: '',
+			tagFilters: {
+				op: 'AND',
+				items: [],
+			},
+			groupBy: [],
+			disabled: false,
+			expression: queryLabel,
+		};
+		setMetricQueries({ ...queries });
+	}, [getNextQueryLabel, metricQueries, setMetricQueries]);
+
+	const addFormula = useCallback(async () => {
+		if (Object.keys(formulaQueries).length === 1) {
+			notification.error({
+				message:
+					'can not add formula. Only one formula is supported in the alert rule.',
+			});
+			return;
+		}
+
+		const queryLabel = getNextQueryLabel();
+
+		const formulas = formulaQueries;
+		formulas[queryLabel] = {
+			queryName: queryLabel,
+			name: queryLabel,
+			formulaOnly: true,
+			expression: 'A',
+			disabled: false,
+		};
+
+		setFormulaQueries({ ...formulas });
+	}, [getNextQueryLabel, formulaQueries, setFormulaQueries]);
+
+	const renderAddons = (): JSX.Element => {
+		return (
+			<ButtonContainer>
+				<QueryButton onClick={addMetricQuery} icon={<PlusOutlined />}>
+					{t('button_query')}
+				</QueryButton>
+				<QueryButton onClick={addFormula} icon={<PlusOutlined />}>
+					{t('button_formula')}
+				</QueryButton>
+			</ButtonContainer>
+		);
+	};
+
+	const renderPromqlUI = (): JSX.Element => {
+		return (
+			<PromqlSection promQueries={promQueries} setPromQueries={setPromQueries} />
+		);
+	};
+
+	const renderMetricUI = (): JSX.Element => {
```

Comment:
Same here also can we wrap this one in useMemo

Reply:
will handle in future releases, as typescript is throwing me off with use of useMemo